atomic_reactor/plugins/prepub_flatpak_create_oci.py

```diff
@@ -0,0 +1,334 @@
+        out_fileobj.close()
+
+        return outfile
+
+    def _export_filesystem(self):
+        image = self.workflow.image
+        self.log.info("Creating temporary docker container")
+        container_dict = self.tasker.d.create_container(image)
+        container_id = container_dict['Id']
+
+        try:
+            return self._export_container(container_id)
+        finally:
+            self.log.info("Cleaning up docker container")
+            self.tasker.d.remove_container(container_id)
+
+    def _create_runtime_oci(self, tarfile, outfile):
+        info = self.source.flatpak_json
+
+        builddir = os.path.join(self.workflow.source.workdir, "build")
+        os.mkdir(builddir)
+
+        repo = os.path.join(self.workflow.source.workdir, "repo")
+        subprocess.check_call(['ostree', 'init', '--mode=archive-z2', '--repo', repo])
+
+        runtime_id = info['runtime']
+        runtime_version = info['runtime-version']
+
+        args = {
+            'runtime_id': runtime_id,
+            'arch': get_arch(),
+            'runtime_version': runtime_version
+        }
+
+        METADATA_TEMPLATE = dedent("""\
+            [Runtime]
+            name={runtime_id}
+            runtime={runtime_id}/{arch}/{runtime_version}
+            sdk={runtime_id}/{arch}/{runtime_version}
+
+            [Environment]
+            LD_LIBRARY_PATH=/app/lib64:/app/lib
+            GI_TYPELIB_PATH=/app/lib64/girepository-1.0
+            """)
+
+        with open(os.path.join(builddir, 'metadata'), 'w') as f:
+            f.write(METADATA_TEMPLATE.format(**args))
+
+        runtime_ref = 'runtime/{runtime_id}/{arch}/{runtime_version}'.format(**args)
+
+        subprocess.check_call(['ostree', 'commit',
+                               '--repo', repo, '--owner-uid=0',
+                               '--owner-gid=0', '--no-xattrs',
+                               '--branch', runtime_ref,
+                               '-s', 'build of ' + runtime_ref,
+                               '--tree=tar=' + tarfile,
+                               '--tree=dir=' + builddir])
+        subprocess.check_call(['ostree', 'summary', '-u', '--repo', repo])
+
+        subprocess.check_call(['flatpak', 'build-bundle', repo,
+                               '--oci', '--runtime',
+                               outfile, runtime_id, runtime_version])
+
+        return runtime_ref
+
+    def _create_app_oci(self, tarfile, outfile):
+        info = self.source.flatpak_json
+        app_id = info['id']
+
+        runtime_id = info['runtime']
+        runtime_version = info['runtime-version']
+
+        builddir = os.path.join(self.workflow.source.workdir, "build")
+        os.mkdir(builddir)
+
+        repo = os.path.join(self.workflow.source.workdir, "repo")
+
+        subprocess.check_call(['flatpak', 'build-init',
+                               builddir, app_id, runtime_id, runtime_id, runtime_version])
+        subprocess.check_call(['tar', 'xCf', builddir, tarfile])
```

Comment:
The question is performance - from experience writing this plugin _compression_ with tarfile is very slow - the reason for the dance with streaming to gzip when extracting and munging the filesystem. But what about decompression? Testing decompression with an actual 91-megabyte filesystem tarball:

```
$ time tar xfz filesystem.tar.gz
real	0m2.369s
user	0m1.988s
sys	0m1.038s

python:
    import sys, tarfile;
    tf = tarfile.open(sys.argv[1]);
    tf.extractall();
    tf.close();
real	0m4.803s
user	0m3.577s
sys	0m1.105s
```

Which is sort of indeterminate - I don't like slowing down a build of a large module by 3 seconds, but it's not that bad. For comparison, compression is a much bigger deal:

```
$ time tar cfz foo.tar.gz file
real	0m12.276s
user	0m12.137s
sys	0m0.498s

python:
     import tarfile, sys;
     tf = tarfile.open(sys.argv[1], "w:gz");
     tf.add(sys.argv[2]);
     tf.close();
real	0m44.857s
user	0m44.317s
sys	0m0.406s
```

On the other hand, when tarring up the resulting OCI image - a few large files with no compression - the Python vs. tar difference is irrelevant:

```
$ time tar cf foo.tar org.fedoraproject.Platform
real	0m0.122s
user	0m0.005s
sys	0m0.116s


Python:
real	0m0.157s
user	0m0.038s
sys	0m0.115s
```

(The conclusion from this might be to drop the compression for the intermediate filesystem tarball, but laptop SSD IO is not the same as IO on the build system, so hard to know...)

Long story short, my plan is:
 * Drop all other usages of subprocess.check_call(['tar', ...])
 * But keep the one you commented in with the comment 'With gzip'ed tarball, tar is several seconds faster than tarfile.extractall'